src/dir.c

```diff
@@ -1212,52 +1226,131 @@ check_tablespace_mapping(pgBackup *backup, bool incremental, bool *tblspaces_are
+	 * of us to drop already existing content of "/somedirectory1".
+	 *
+	 * TODO: Ideally in case of incremental restore we must also
+	 * drop the "/somedirectory" directory first, but currently
+	 * we don`t do that.
+	 */
+
 	/* 2 - all linked directories must be empty */
 	for (i = 0; i < parray_num(links); i++)
 	{
 		pgFile	   *link = (pgFile *) parray_get(links, i);
 		const char *linked_path = link->linked;
 		TablespaceListCell *cell;
+		bool remapped = false;
 
 		for (cell = tablespace_dirs.head; cell; cell = cell->next)
 			if (strcmp(link->linked, cell->old_dir) == 0)
 			{
 				linked_path = cell->new_dir;
+				remapped = true;
 				break;
 			}
 
 		if (!is_absolute_path(linked_path))
-			elog(ERROR, "tablespace directory is not an absolute path: %s\n",
+			elog(ERROR, "Tablespace directory is not an absolute path: %s\n",
 				 linked_path);
 
 		if (!dir_is_empty(linked_path, FIO_DB_HOST))
 		{
+
 			if (!incremental)
-				elog(ERROR, "restore tablespace destination is not empty: \"%s\"",
-					 linked_path);
-			*tblspaces_are_empty = false;
+				elog(ERROR, "Restore tablespace destination is not empty: \"%s\"", linked_path);
+
+			else if (remapped && !force)
+				elog(ERROR, "Remapped tablespace destination is not empty: \"%s\". "
+							"Use \"--force\" flag if you want to automatically clean up the "
+							"content of new tablespace destination",
+						linked_path);
+
+			else if (pgdata_is_empty && !force)
+				elog(ERROR, "PGDATA is empty, but tablespace destination is not: \"%s\". "
+							"Use \"--force\" flag is you want to automatically clean up the "
+							"content of tablespace destination",
+						linked_path);
+
+			/*
+			 * TODO: compile the list of tblspc Oids to delete later,
+			 * similar to what we do with database_map.
+			 */
+			else if (force && (pgdata_is_empty || remapped))
+			{
+				// TODO: move this stuff to separate function.
+				char	fullpath[MAXPGPATH];
+				int i;
+				parray *files = parray_new();
+
+				elog(WARNING, "Cleaning up the content of %s directory: \"%s\"",
+							remapped ? "remapped tablespace" : "tablespace",
+							linked_path);
+
+				fio_list_dir(files, linked_path, false, false, false, false, false, 0);
+
+				/* delete leaf node first */
+				parray_qsort(files, pgFileCompareRelPathWithExternalDesc);
+
+				for (i = 0; i < parray_num(files); i++)
+				{
+					pgFile	   *file = (pgFile *) parray_get(files, i);
+
+					join_path_components(fullpath, linked_path, file->rel_path);
+
+					fio_delete(file->mode, fullpath, FIO_DB_HOST);
+					elog(VERBOSE, "Deleted file \"%s\"", fullpath);
+				}
+
+				parray_walk(files, pgFileFree);
+				parray_free(files);
+				continue;
+			}
+
+			tblspaces_are_empty = false;
 		}
 	}
 
 	free(tmp_file);
 	parray_walk(links, pgFileFree);
 	parray_free(links);
+
+	if (tblspaces_are_empty)
+		return 1;
+	return 2;
 }
 
+/* Make it consistent with check_tablespace_mapping */
```

Comment:
I don't understand the comment. Is it TODO?

Reply:
Yes.